Register CPU/CUDA fuser dynamically

zdevito opened the pull request:
This avoids a bunch of conditional compilation logic

zdevito commented:
@pytorchbot retest this please.

zou3519 commented:
Taking a look. cc @mruberry too (I think you wrote most of the conditional compilation logic originally?)

zou3519 commented on tools/build_variables.py:
```diff
@@ -99,6 +99,14 @@
+    "torch/csrc/jit/fuser/codegen.cpp",
+    "torch/csrc/jit/fuser/fallback.cpp",
+    "torch/csrc/jit/fuser/cpu/fused_kernel.cpp",
+    "torch/csrc/jit/fuser/cpu/dynamic_library_unix.cpp",
```
Is this list for libtorch only? Do we support libtorch on windows? (we're including dynamic_library_unix here so it sounds like no but just checking)

zdevito replied:
This is the buck build. No windows.

zou3519 commented on tools/build_variables.py:
```diff
@@ -171,8 +176,6 @@ def torch_vars():
             "-Icaffe2/torch/csrc",
             "-Icaffe2/torch/csrc/nn",
             "-Icaffe2/torch/lib",
-            "-DUSE_CPU_FUSER_FBCODE=1",
```
Are we still disabling the CPU Fuser in fbcode? I'm not sure where that logic lives... (hopefully it is in a buck file somewhere)

zdevito replied:
I copied the rules for what was there. According to this code it is enabled in fbcode.

zou3519 reviewed (approved):
lgtm, but the windows build is complaining about a line that doesn't exist anymore:

```
19:50:32 C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-build\torch/csrc/jit/fuser/cuda/resource_strings.h(55): fatal error C1020: unexpected #endif
```
which is really strange

zdevito commented:
I know, I am not sure what is going on there.

mruberry commented:
Thanks for cc, it's great to see the build logic generalized and simplified like this.

zdevito committed:
Register CPU/CUDA fuser dynamically (#15887)

Summary:
This avoids a bunch of conditional compilation logic
Pull Request resolved: https://github.com/pytorch/pytorch/pull/15887

Differential Revision: D13613239

fbshipit-source-id: b3f5e6c0c52cf737b2996405f76005dad48a8ae4
